records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,232 @@
+import json
+import time
+from contextlib import contextmanager
+from records_mover.url.base import BaseDirectoryUrl
+import logging
+from typing import Iterator, Tuple, Union, Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from googleapiclient.discovery import _TransferJobConfig
+    import google.auth.credentials
+    from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+    from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopyOptimizer:
+    # TODO: Figure out where this should be called from
+    # TODO: Document this class
+
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        # TODO: Implement this and swap out call
+        if loc.scheme == 's3' and other_loc.scheme == 'gs':
+            from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, GCSDirectoryUrl)
+            return self.copy_via_gcp_data_transfer(loc, other_loc)
+        else:
+            logger.info(f"No strategy to optimize copy from {loc} to {other_loc}")
+            return False
+
+    def copy_via_gcp_data_transfer(self,
+                                   loc: 'S3DirectoryUrl',
+                                   other_loc: 'GCSDirectoryUrl') -> bool:
+        import googleapiclient
+
+        # https://cloud.google.com/storage-transfer/docs/create-manage-transfer-program#python
+
+        if loc.key != other_loc.blob:
+            logger.warning("S3 directory does not match GCS directory - "
+                           "cannot copy S3 bucket using Google Storage Transfer.  "
+                           "Falling back to a slower method of bucket copy.")
+            return False
+
+        gcp_credentials = other_loc.credentials
+        storagetransfer = googleapiclient.discovery.build('storagetransfer', 'v1',
+                                                          credentials=gcp_credentials)
+        description = "records-mover one-time job"
+        # TODO
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker